### PR TITLE
Knn docs

### PR DESCRIPTION
**Pull request by azevaykin**
### Changelog entry 

...

### Changelog category 

* Documentation (changelog entry is not required)

### Additional information

...

### CODE REVIEW

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/_includes/index/list.md [hunk 1]
```diff
@@ -1,13 +1,14 @@
+* [DateTime](../../datetime.md)
+* [Digest](../../digest.md)
+* [Histogram](../../histogram.md)
 * [Hyperscan](../../hyperscan.md)
+* [Ip](../../ip.md)
+* [Knn](../../knn.md)
+* [Math](../../math.md)
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/_includes/index/list.md:7]: The previous order looks like by descending popularity, and you're changing it to lexicographic order. There might be a discussion on which one is better, but in any case, this change is unrelated to KNN.

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
+## Introduction
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:1]:
```suggestion
# KNN
## Introduction
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+[Nearest Neighbor search](https://en.wikipedia.org/wiki/Nearest_neighbor_search) (NN) is an optimization task that consists in finding the closest point (or set of points) in a given dataset to a given query point. Closeness can be defined in terms of distance or similarity metrics.
+A generalization of the NN problem is the k-NN problem, where it's required to find the k nearest points to the query point. This can be useful in various applications such as image classification, recommendation systems, and more.
+https://en.wikipedia.org/wiki/Nearest_neighbor_search
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:5]:
```suggestion
[Nearest Neighbor search](https://en.wikipedia.org/wiki/Nearest_neighbor_search) (NN) is an optimization task that consists of finding the closest point in a given dataset to a given query point. Closeness can be defined in terms of distance or similarity metrics.
A generalization of the NN problem is the [k-NN problem](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm), where it's required to find the `k` nearest points to the query point. This can be useful in various applications such as image classification, recommendation systems, etc.
https://en.wikipedia.org/wiki/Nearest_neighbor_search
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+The solution of the problem is divided into two major subclasses of methods: exact and approximate. In this document, we will talk about the exact brute force method.
+
+The method is based on calculating the distance from the query point to every other point in the database. This algorithm, also known as the naive approach, has a complexity of `O(dn)`, where `n` is the number of points in the dataset, and `d` is its dimension.
+
+The advantage of the method is that there is no need for additional data structures, such as specialized vector indexes.
+The disadvantage is the need for a complete data search. But this disadvantage is insignificant in cases where data has been pre-filtered, for example, by user ID.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:12]:
```suggestion
The k-NN problem solution is divided into two major subclasses of methods: exact and approximate. In this document, we will talk about the exact brute force method.

This method is based on calculating the distance from the query point to every other point in the database. This algorithm, also known as the naive approach, has a complexity of `O(dn)`, where `n` is the number of points in the dataset, and `d` is its dimension.

The advantage of the method is that there is no need for additional data structures, such as specialized vector indexes.
The disadvantage is the need for a complete data search. But this disadvantage is insignificant in cases where data has been pre-filtered, for example, by user ID.
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Example:
+
+```sql
+SELECT id, fact, embedding FROM Facts
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:17]:
```suggestion
DECLARE $TargetEmbedding AS List<Float>;
SELECT id, fact, embedding FROM Facts
```

Comment (blinkov) [reply ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:17]: Otherwise, the query won't be valid.

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+WHERE user="Williams"
+ORDER BY Knn::CosineDistance(Knn::FromBinaryString(embedding), $TargetEmbedding)
+LIMIT 10
+```
+
+## Data type
+
+In mathematics, a vector of real numbers is used to store points.
+In YDB, for calculations the `List<Float>` data type is used, and data will be stored in the string data type `String`.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:26]:
```suggestion
In {{ ydb-short-name }}, for calculations the `List<Float>` data type is used, and data will be stored in the `String` data type.
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+Example:
+
+```sql
+SELECT id, fact, embedding FROM Facts
+WHERE user="Williams"
+ORDER BY Knn::CosineDistance(Knn::FromBinaryString(embedding), $TargetEmbedding)
+LIMIT 10
+```
+
+## Data type
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:23]:
```suggestion
## Data types
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+* euclidean distance `EuclideanDistance' (the root of the sum of squares of coordinate differences)
+* cosine distance `CosineDistance' (1 - cosine similarity)
+
+#### Function signature
+
+```sql
+Knn::InnerProductSimilarity(List<Float>?, List<Float>?)->Float?
+Knn::CosineSimilarity(List<Float>?, List<Float>?)->Float?
+Knn::EuclideanDistance(List<Float>?, List<Float>?)->Float?
+Knn::CosineDistance(List<Float>?, List<Float>?)->Float?
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:51]:
```suggestion
Knn::InnerProductSimilarity(List<Float>{Flags:AutoMap}, List<Float>{Flags:AutoMap})->Float?
Knn::CosineSimilarity(List<Float>{Flags:AutoMap}, List<Float>{Flags:AutoMap})->Float?
Knn::EuclideanDistance(List<Float>{Flags:AutoMap}, List<Float>{Flags:AutoMap})->Float?
Knn::CosineDistance(List<Float>{Flags:AutoMap}, List<Float>{Flags:AutoMap})->Float?
```
as far as I can tell that's what it ended up being in #2519, but please double-check

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+In YDB, for calculations the `List<Float>` data type is used, and data will be stored in the string data type `String`.
+
+## Functions
+
+Vector functions are implemented as user-defined functions (UDF) in the `Knn` module.
+
+### Distance and similarity functions
+
+The distance and similarity functions take two lists of real numbers as input and return the distance/similarity between them.
+Distance functions return small values for close vectors, similarity functions return large values for close vectors. This should be taken into account in the sorting order.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:35]:
```suggestion
The distance and similarity functions take two lists of real numbers as input and return the distance/similarity between them.

{% note info %}

Distance functions return small values for close vectors, while similarity functions return large values for close vectors. This should be taken into account when defining the sorting order.

{% endnote %}
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+Similarity functions:
+* inner product `InnerProductSimilarity' (sum of products of coordinates)
+* cosine similarity `CosineSimilarity' (inner product / vector lengths)
+
+Distance functions:
+* euclidean distance `EuclideanDistance' (the root of the sum of squares of coordinate differences)
+* cosine distance `CosineDistance' (1 - cosine similarity)
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:43]:
```suggestion
Similarity functions:
* inner product `InnerProductSimilarity` (sum of products of coordinates)
* cosine similarity `CosineSimilarity` (inner product / vector lengths)

Distance functions:
* euclidean distance `EuclideanDistance` (the root of the sum of squares of coordinate differences)
* cosine distance `CosineDistance` (1 - cosine similarity)
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+#### Function signature
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:45]:
```suggestion
#### Function signatures
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+```sql
+Knn::InnerProductSimilarity(List<Float>?, List<Float>?)->Float?
+Knn::CosineSimilarity(List<Float>?, List<Float>?)->Float?
+Knn::EuclideanDistance(List<Float>?, List<Float>?)->Float?
+Knn::CosineDistance(List<Float>?, List<Float>?)->Float?
+```
+
+In case of error, the functions return NULL.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:54]:
```suggestion
In case of an error, these functions return `NULL`.
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+### Functions for converting a vector to a binary representation
+
+Conversion functions are needed to serialize the vector set into an internal binary representation and vice versa.
+The binary representation of the vector will be stored in the YDB database in the `String` type.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:59]:
```suggestion
### Functions for converting between vector and binary representations

Conversion functions are needed to serialize the vector set into an internal binary representation and vice versa.
The binary representation of a vector can be persisted in a {{ ydb-short-name }}  table column of the `String` type.
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+#### Function signature
+
+```sql
+Knn::ToBinaryString(List<Float>)->String
+Knn::FromBinaryString(String)->List<Float>?
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:65]: I think FromBinaryString could use TAutoMap too, you might consider updating https://github.com/azevaykin/ydb/blob/f21ee1198d8ff0def282b9feba6f0abcc77d9e06/ydb/library/yql/udfs/common/knn/knn.cpp#L20

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+In case of deserialization error, the `FromBinaryString` function return NULL.
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:68]:
```suggestion
#### Function signatures

```sql
Knn::ToBinaryString(List<Float>{Flags:AutoMap})->String
Knn::FromBinaryString(String)->List<Float>?
```

In case of a deserialization error, the `FromBinaryString` function returns `NULL`.
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+## Examples
+
+### Creating a table
+
+```sql
+CREATE TABLE Facts (
+    id Uint64,        // Id of fact
+    user String,      // User name
+    fact String,      // Human readable description of user fact
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:78]:
```suggestion
    fact String,      // Human-readable description of a user fact
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    embedding String, // Binary representation of embedding vector (result of Knn::ToBinaryString)
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:79]:
```suggestion
    embedding String, // Binary representation of embedding vector (a result of Knn::ToBinaryString)
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+The advantage of the method is that there is no need for additional data structures, such as specialized vector indexes.
+The disadvantage is the need for a complete data search. But this disadvantage is insignificant in cases where data has been pre-filtered, for example, by user ID.
+
+Example:
+
+```sql
+SELECT id, fact, embedding FROM Facts
+WHERE user="Williams"
+ORDER BY Knn::CosineDistance(Knn::FromBinaryString(embedding), $TargetEmbedding)
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:19]:
```suggestion
ORDER BY Knn::CosineDistance(
    Knn::FromBinaryString(embedding),
    $TargetEmbedding
)
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+### Exact vector search
+
+```sql
+$TargetEmbedding = [1.2f, 2.3f, 3.4f, 4.5f]
+
+SELECT * FROM Facts
+WHERE user="Williams"
+ORDER BY Knn::CosineDistance(Knn::FromBinaryString(embedding), $TargetEmbedding)
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:98]:
```suggestion
ORDER BY Knn::CosineDistance(
    Knn::FromBinaryString(embedding),
    $TargetEmbedding
)
```

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/toc_base.yaml [hunk 20]
```diff
@@ -1,16 +1,17 @@
 items:
 - name: Overview
   href: index.md
+- { name: DateTime,               href: datetime.md }
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/udf/list/toc_base.yaml:4]: the same comment about order as above

**Review by blinkov — CHANGES_REQUESTED** [review]

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/_includes/index/list.md (hunk 1 above)
Comment (azevaykin) [reply ydb/docs/en/core/yql/reference/yql-core/udf/list/_includes/index/list.md]: 39ec11af2cfeae6592e34d41edd9a883b846823c

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/toc_base.yaml (hunk 20 above)
Comment (azevaykin) [reply ydb/docs/en/core/yql/reference/yql-core/udf/list/toc_base.yaml]: 39ec11af2cfeae6592e34d41edd9a883b846823c

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md (hunk 14 above)
Comment (azevaykin) [reply ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:74]: https://github.com/ydb-platform/ydb/pull/2795

File: ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md (hunk 5 above)
Comment (azevaykin) [reply ydb/docs/en/core/yql/reference/yql-core/udf/list/knn.md:20]: eb58508fbc525ff955239f046a5c120537268724

**Review by fomichev3000 — DISMISSED** [review]

File: ydb/docs/ru/core/yql/reference/yql-core/udf/list/knn.md [hunk 21]
```diff
@@ -0,0 +1,111 @@
+# KNN
+## Введение
+
+[Поиск ближайшего соседа](https://en.wikipedia.org/wiki/Nearest_neighbor_search) (NN) - это задача оптимизации, заключающаяся в нахождении ближайшей точки (или набора точек) в заданном наборе данных к заданной точке запроса. Близость может быть определена в терминах метрики расстояния или сходства.
+Обобщением задачи NN является задача [k-NN](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm), где от нас требуется найти k ближайших точек к точке запроса. Это может быть полезно в различных приложениях, таких как классификация изображений, рекомендательные системы и многое другое.
+
+Решения задачи k-NN разбивается на два крупных подкласса методов: точные и приближенные. В данном документе речь пойдет о точном походе, а точнее о точном подходе методом грубой силы.
```
[the remark this answers is not in the document]
Comment (blinkov) [reply ydb/docs/ru/core/yql/reference/yql-core/udf/list/knn.md:7]:
```suggestion
Решения задачи k-NN разбивается на два крупных подкласса методов: точные и приближенные. В данном документе речь пойдет о точном подходе, а именно о точном подходе методом грубой силы.
```

**Review by blinkov — APPROVED** [review]
